src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php

```diff
@@ -61,24 +51,21 @@ public function check(ResourceControllerEvent $event): void
 
         Assert::isInstanceOf($order, OrderInterface::class);
 
-        // we create a new promotion collection and remove them from cart
-        // so we can verify with original conditions (without the price being applied before check)
-
-        $promotions = $order->getPromotions()->toArray();
+        $previouslyAppliedPromotionsCodeToNameMap = $this->createPromotionCodeToNameMap($order);
```

Comment:
Did you think about just cloning order's promotion collection instead?
Something like this should work: `$previousPromotions = new ArrayCollection($order->getPromotions()->toArray());`
And after processing just loop through `$order->getPromotions()` and check if `$previousPromotions->contains($promotion)`.
Not necessarily better, just a matter of taste I guess.